finality-aleph/src/validator_network/service.rs

```diff
@@ -150,76 +154,122 @@ impl<D: Data, A: Data + Debug, ND: Dialer<A>, NL: Listener> Service<D, A, ND, NL
-                        Added => info!(target: "validator-network", "New incoming connection for peer {}.", peer_id),
-                        Replaced => info!(target: "validator-network", "Replaced incoming connection for peer {}.", peer_id),
-                    }
-                },
-                // received information from a spawned worker managing an outgoing connection
+                // received information from a spawned worker managing a connection
                 // check if we still want to be connected to the peer, and if so, spawn a new worker or actually add proper connection
-                Some((peer_id, maybe_data_for_network, _)) = outgoing_workers.next() => {
-                    use AddResult::*;
-                    if let Some(addresses) = self.manager.peer_addresses(&peer_id) {
-                        match maybe_data_for_network {
-                            Some(data_for_network) => match self.manager.add_outgoing(peer_id.clone(), data_for_network) {
-                                Uninterested => warn!(target: "validator-network", "We connected to peer {} for unknown reasons.", peer_id),
-                                Added => info!(target: "validator-network", "New outgoing connection to peer {}.", peer_id),
-                                Replaced => info!(target: "validator-network", "Replaced outgoing connection to peer {}.", peer_id),
-                            },
-                            None => self.spawn_new_outgoing(peer_id, addresses, outgoing_result_for_parent.clone()),
+                Some((peer_id, maybe_data_for_network, connection_type)) = worker_results.next() => {
+                    use ConnectionType::*;
+                    let spawn_new_legacy_connection = match connection_type {
+                        LegacyIncoming => {
+                            self.manager.remove_peer(&peer_id);
+                            self.legacy_connected.insert(peer_id.clone())
+                        },
+                        LegacyOutgoing => {
+                            self.manager.remove_peer(&peer_id);
+                            self.legacy_connected.insert(peer_id.clone());
+                            false
                         }
+                        New => {
+                            // We always return New if connecting fails, so only New+Some means we
+                            // actually negotiated the new protocol, otherwise we should stay
+                            // with our previous guess.
+                            if maybe_data_for_network.is_some() {
+                                self.legacy_connected.remove(&peer_id);
+                            }
+                            false
+                        },
                     };
+                    if spawn_new_legacy_connection {
+                        match self.legacy_manager.peer_addresses(&peer_id) {
```

Comment:
Maybe this could be moved and changed to
```
if self.legacy_connected.insert(peer_id.clone()) {
    Some(addresses) => self.spawn_new_outgoing(peer_id.clone(), addresses, result_for_parent.clone()),
    None => {
        self.legacy_connected.remove(&peer_id);
    },
}
```
This match with setting `spawn_new_legacy_connection` seems unnecessary